File: Hash table/Subarray_Sum_Equals_K.py
"""Given an array of integers and an integer k, you need to find 
the total number of continuous subarrays whose sum equals to k."""

class Solution(object):
    def subarraySum(self, nums, k):
        res =0 
        d = dict()
        sm = 0
        d[0] = 1
        for i in range(len(nums)):
            sm += nums[i]
            if sm-k in d:
                res += d[sm-k]
            if sm in d:
                d[sm] += 1
            else:
                d[sm] = 1
        return res 

File: Hash table/SubarraySumEqualsK.cpp
//problem Link:- https://leetcode.com/problems/subarray-sum-equals-k/

#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int sm = 0;
        map<int,int> mp;
        mp[0] = 1;
        int res = 0;
        for(int i=0;i<nums.size();i++){
            sm += nums[i];
            if (mp.find(sm-k) != mp.end()){
                res += mp[sm-k];
            }
            mp[sm]++;
        }
        return res;
    }
};